README
--------------------------------------------------------------------
libbriteblox version 1.1
--------------------------------------------------------------------

libbriteblox - A library (using libusb) to talk to BRITEBLOX's UART/FIFO chips
including the popular bitbang mode.

The following chips are supported:
- FT4232H / FT2232H
- FT232R  / FT245R
- FT2232L / FT2232D / FT2232C
- FT232BM / FT245BM (and the BL/BQ variants)
- FT8U232AM / FT8U245AM

libbriteblox requires libusb 1.x.

The AUTHORS file contains a list of all the people
that made libbriteblox possible what it is today.

Changes
-------
* Fix FT232H eeprom suspend pulldown setting (Davide Michelizza)
* Fix FT232H eeprom user area size (Davide Michelizza)
* Improved mingw build (Paul Fertser and Michel Zou)
* C++ wrapper: Get/set functions for USB timeouts (Jochen Sprickerhof)
* Partial support for FT230X (Nathael Pajani)
* New API function: briteblox_eeprom_set_strings() (Nathael Pajani)
* Prevent possible segfault in briteblox_eeprom_decode() (Nathael Pajani)
* Save device release number in eeprom (Jarkko Sonninen)
* Fix "self powered" eeprom flag (Jarkko Sonninen)
* Improved python wrapper (Michel Zou)
* Many buildsystem improvements (Michel Zou and Mike Frysinger)
* See the git history for more changes and fixes

You'll find the newest version of libbriteblox at:
http://www.intra2net.com/en/developer/libbriteblox


Quick start
-----------
mkdir build
cd build

cmake -DCMAKE_INSTALL_PREFIX="/usr" ../
make
make install

More verbose build instructions are in "README.build"

--------------------------------------------------------------------
www.intra2net.com                             2003-2014 Intra2net AG
--------------------------------------------------------------------


ChangeLog
New in 1.1 - 2014-02-05
-----------------------
* Fix FT232H eeprom suspend pulldown setting (Davide Michelizza)
* Fix FT232H eeprom user area size (Davide Michelizza)
* Improved mingw build (Paul Fertser and Michel Zou)
* C++ wrapper: Get/set functions for USB timeouts (Jochen Sprickerhof)
* Partial support for FT230X (Nathael Pajani)
* New API function: briteblox_eeprom_set_strings() (Nathael Pajani)
* Prevent possible segfault in briteblox_eeprom_decode() (Nathael Pajani)
* Save device release number in eeprom (Jarkko Sonninen)
* Fix "self powered" eeprom flag (Jarkko Sonninen)
* Improved python wrapper (Michel Zou)
* Many buildsystem improvements (Michel Zou and Mike Frysinger)
* See the git history for more changes and fixes

New in 1.0 - 2013-01-29
-----------------------
* Ported to libusb 1.x (initial work by Jie Zhang)
* Many eeprom handling improvements (Uwe Bonnes, Anders Larsen)
* Renamed pkconfig, library .so etc. files to "libbriteblox1" (Intra2net)
* briteblox_eeprom is part of libbriteblox now (Intra2net)

* New baudrate calculation code + unit tests (Uwe Bonnes and Intra2net)
* Improved python bindings including python3 support (Michel Zou)
* Switched completely to cmake build system (Intra2net)
* cmake: Easy libbriteblox discovery via find_package() (Michel Zou)
* eeprom handling now done via get()/set() functions (Uwe Bonnes)
* C++ wrapper: Fixed use-after-free in List::find_all() (Intra2net)
* Documentation updates (Xiaofan Chen)
* See the git history for more changes and fixes

New in 0.20 - 2012-03-19
------------------------
* Support for FT232H (Uwe Bonnes)
* Fixed install location of header files (Uwe Bonnes and Intra2net)
* Backported serial_test tool from libbriteblox 1.x (Uwe Bonnes)

New in 0.19 - 2011-05-23
------------------------
* Make kernel driver detach configurable (Thomas Klose)
* Correct briteblox_poll_modem_status() result code (Tom Saunders)
* cmake build system improvements (Evgeny Sinelnikov)
* Fix uninitialized memory access in async mode (Intra2net)
* Support for FT232R eeprom features (Hermann Kraus)
* Fix size returned by briteblox_read_data (Hermann Kraus)
* C++ wrapper: Fix infinite recursion in set_bitmode (Intra2net)
* Improvements to the python wrapper (Flynn Marquardt and Chris Zeh)

New in 0.18 - 2010-06-25
------------------------
* Add briteblox_eeprom_free() to free allocated memory in eeprom (Wilfried Holzke)
* More generic error message for the BRITEBLOX kernel driver (Intra2net)
* Honor CPPFLAGS in python wrapper build (Alexander Lehmann)
* cmake: Fix package creation on 32-bit machines (Uwe Bonnes)
* Fix swig argument constraints (Intra2net)
* Don't segfault if device is closed or briteblox context is invalid (Intra2net)
* Ability to disable build of examples / documentation (Mike Frysinger and Intra2net)
* Fix typo in python wrapper build (Mike Frysinger)
* Autoconf build system improvements (Mike Frysinger)

New in 0.17 - 2009-12-19
------------------------
* C++ wrapper: Reduced code duplication and small other changes (Intra2net)
* Deprecated old briteblox_enable_bitbang() function (Intra2net)
* New briteblox_usb_open_desc_index() function (Intra2net)
* Added baud rate test example code (Intra2net)
* New serial input example code (Jim Paris)
* Fix modem status byte filtering for USB high speed chips (Intra2net and Jim Paris)
* Add bitmode for synchronous fifo in FT2232H (Uwe Bonnes)
* Fix usb_set_configuration() call on Windows 64 (NIL)
* Fix usb index in briteblox_convert_baudrate() for FT2232H/FT4232H chips (Thimo Eichstaedt)
* Set initial baudrate on correct interface instead of always the first one (Thimo Eichstaedt)
* Call usb_set_configuration() on Windows only (Uwe Bonnes)
* 64 bit and other buildsystem fixes (Uwe Bonnes)
* Don't build --with-async-mode w/ libusb-compat-0.1 (Clifford Wolf)
* Functions for read/write of a single eeprom location (Oleg Seiljus)
* Protect against double close of usb device (Nathan Fraser)
* Fix out-of-tree-build in python wrapper (Aurelien Jarno)
* Autoconf and doxygen cleanup (Jim Paris)

New in 0.16 - 2009-05-08
------------------------
* C++ wrapper: Reopen the device after calling get_strings() in Context::open() (Marek Vavruša and Intra2net)
* C++ wrapper: Fixed an inheritance problem (Marek Vavruša and Intra2net)
* C++ wrapper: Relicensed under GPLv2 + linking exception (Marek Vavruša and Intra2net)
* Support for FT2232H and FT4232H (David Challis, Alex Harford and Intra2net)
* Support for mingw cross compile (Uwe Bonnes)
* Python bindings and minor autoconf cleanup (Tarek Heiland)
* Code cleanup in various places (Intra2net)
* Fixed briteblox_read_chipid in some cases (Matthias Richter)
* eeprom decode function and small cleanups (Marius Kintel)
* cmake system improvements (Marius Kintel and Intra2net)
* Fix compilation in -ansi -pedantic mode (Matthias Janke)

New in 0.15 - 2008-12-19
------------------------

src/ftdi.h
/***************************************************************************
                          briteblox.h  -  description
                             -------------------
    begin                : Fri Apr 4 2003
    copyright            : (C) 2003-2014 by Intra2net AG and the libbriteblox developers
    email                : opensource@intra2net.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Lesser General Public License           *
 *   version 2.1 as published by the Free Software Foundation;             *
 *                                                                         *
 ***************************************************************************/

#ifndef __libbriteblox_h__
#define __libbriteblox_h__

#include <stdint.h>
#include <sys/time.h>

/* 'interface' might be defined as a macro on Windows, so we need to
 * undefine it so as not to break the current libbriteblox API, because
 * struct briteblox_context has an 'interface' member
 * As this can be problematic if you include windows.h after briteblox.h
 * in your sources, we force windows.h to be included first. */
#if defined(_WIN32) || defined(__CYGWIN__) || defined(_WIN32_WCE)
#include <windows.h>
#if defined(interface)
#undef interface
#endif
#endif

/** BRITEBLOX chip type */
enum briteblox_chip_type
{
    TYPE_AM=0,
    TYPE_BM=1,
    TYPE_2232C=2,
    TYPE_R=3,
    TYPE_2232H=4,
    TYPE_4232H=5,
    TYPE_232H=6,
    TYPE_230X=7,
};
/** Parity mode for briteblox_set_line_property() */
enum briteblox_parity_type { NONE=0, ODD=1, EVEN=2, MARK=3, SPACE=4 };
/** Number of stop bits for briteblox_set_line_property() */
enum briteblox_stopbits_type { STOP_BIT_1=0, STOP_BIT_15=1, STOP_BIT_2=2 };
/** Number of bits for briteblox_set_line_property() */
enum briteblox_bits_type { BITS_7=7, BITS_8=8 };
/** Break type for briteblox_set_line_property2() */
enum briteblox_break_type { BREAK_OFF=0, BREAK_ON=1 };

/** MPSSE bitbang modes */
enum briteblox_mpsse_mode
{
    BITMODE_RESET  = 0x00,    /**< switch off bitbang mode, back to regular serial/FIFO */
    BITMODE_BITBANG= 0x01,    /**< classical asynchronous bitbang mode, introduced with B-type chips */
    BITMODE_MPSSE  = 0x02,    /**< MPSSE mode, available on 2232x chips */
    BITMODE_SYNCBB = 0x04,    /**< synchronous bitbang mode, available on 2232x and R-type chips  */
    BITMODE_MCU    = 0x08,    /**< MCU Host Bus Emulation mode, available on 2232x chips */
    /* CPU-style fifo mode gets set via EEPROM */
    BITMODE_OPTO   = 0x10,    /**< Fast Opto-Isolated Serial Interface Mode, available on 2232x chips  */
    BITMODE_CBUS   = 0x20,    /**< Bitbang on CBUS pins of R-type chips, configure in EEPROM before */
    BITMODE_SYNCFF = 0x40,    /**< Single Channel Synchronous FIFO mode, available on 2232H chips */
    BITMODE_FT1284 = 0x80,    /**< FT1284 mode, available on 232H chips */
};

/** Port interface for chips with multiple interfaces */
enum briteblox_interface
{
    INTERFACE_ANY = 0,
    INTERFACE_A   = 1,
    INTERFACE_B   = 2,
    INTERFACE_C   = 3,
    INTERFACE_D   = 4
};

/** Automatic loading / unloading of kernel modules */
enum briteblox_module_detach_mode
{
    AUTO_DETACH_SIO_MODULE = 0,
    DONT_DETACH_SIO_MODULE = 1
};

/* Shifting commands IN MPSSE Mode*/
#define MPSSE_WRITE_NEG 0x01   /* Write TDI/DO on negative TCK/SK edge*/
#define MPSSE_BITMODE   0x02   /* Write bits, not bytes */
#define MPSSE_READ_NEG  0x04   /* Sample TDO/DI on negative TCK/SK edge */
#define MPSSE_LSB       0x08   /* LSB first */
#define MPSSE_DO_WRITE  0x10   /* Write TDI/DO */
#define MPSSE_DO_READ   0x20   /* Read TDO/DI */
#define MPSSE_WRITE_TMS 0x40   /* Write TMS/CS */

/* BRITEBLOX MPSSE commands */
#define SET_BITS_LOW   0x80
/*BYTE DATA*/
/*BYTE Direction*/
#define SET_BITS_HIGH  0x82
/*BYTE DATA*/
/*BYTE Direction*/
#define GET_BITS_LOW   0x81
#define GET_BITS_HIGH  0x83
#define LOOPBACK_START 0x84
#define LOOPBACK_END   0x85
#define TCK_DIVISOR    0x86
/* H Type specific commands */
#define DIS_DIV_5       0x8a
#define EN_DIV_5        0x8b
#define EN_3_PHASE      0x8c
#define DIS_3_PHASE     0x8d
#define CLK_BITS        0x8e
#define CLK_BYTES       0x8f
#define CLK_WAIT_HIGH   0x94
#define CLK_WAIT_LOW    0x95
#define EN_ADAPTIVE     0x96
#define DIS_ADAPTIVE    0x97
#define CLK_BYTES_OR_HIGH 0x9c
#define CLK_BYTES_OR_LOW  0x0d
/*FT232H specific commands */
#define DRIVE_OPEN_COLLECTOR 0x9e
/* Value Low */
/* Value HIGH */ /*rate is 12000000/((1+value)*2) */
#define DIV_VALUE(rate) (rate > 6000000)?0:((6000000/rate -1) > 0xffff)? 0xffff: (6000000/rate -1)

/* Commands in MPSSE and Host Emulation Mode */
#define SEND_IMMEDIATE 0x87
#define WAIT_ON_HIGH   0x88
#define WAIT_ON_LOW    0x89

/* Commands in Host Emulation Mode */
#define READ_SHORT     0x90
/* Address_Low */
#define READ_EXTENDED  0x91
/* Address High */
/* Address Low  */
#define WRITE_SHORT    0x92
/* Address_Low */
#define WRITE_EXTENDED 0x93
/* Address High */
/* Address Low  */

/* Definitions for flow control */
#define SIO_RESET          0 /* Reset the port */
#define SIO_MODEM_CTRL     1 /* Set the modem control register */
#define SIO_SET_FLOW_CTRL  2 /* Set flow control register */
#define SIO_SET_BAUD_RATE  3 /* Set baud rate */
#define SIO_SET_DATA       4 /* Set the data characteristics of the port */

#define BRITEBLOX_DEVICE_OUT_REQTYPE (LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_RECIPIENT_DEVICE | LIBUSB_ENDPOINT_OUT)
#define BRITEBLOX_DEVICE_IN_REQTYPE (LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_RECIPIENT_DEVICE | LIBUSB_ENDPOINT_IN)

/* Requests */
#define SIO_RESET_REQUEST             SIO_RESET
#define SIO_SET_BAUDRATE_REQUEST      SIO_SET_BAUD_RATE
#define SIO_SET_DATA_REQUEST          SIO_SET_DATA
#define SIO_SET_FLOW_CTRL_REQUEST     SIO_SET_FLOW_CTRL
#define SIO_SET_MODEM_CTRL_REQUEST    SIO_MODEM_CTRL
#define SIO_POLL_MODEM_STATUS_REQUEST 0x05
#define SIO_SET_EVENT_CHAR_REQUEST    0x06
#define SIO_SET_ERROR_CHAR_REQUEST    0x07
#define SIO_SET_LATENCY_TIMER_REQUEST 0x09
#define SIO_GET_LATENCY_TIMER_REQUEST 0x0A
#define SIO_SET_BITMODE_REQUEST       0x0B
#define SIO_READ_PINS_REQUEST         0x0C
#define SIO_READ_EEPROM_REQUEST       0x90
#define SIO_WRITE_EEPROM_REQUEST      0x91
#define SIO_ERASE_EEPROM_REQUEST      0x92


#define SIO_RESET_SIO 0
#define SIO_RESET_PURGE_RX 1
#define SIO_RESET_PURGE_TX 2

#define SIO_DISABLE_FLOW_CTRL 0x0
#define SIO_RTS_CTS_HS (0x1 << 8)
#define SIO_DTR_DSR_HS (0x2 << 8)
#define SIO_XON_XOFF_HS (0x4 << 8)

#define SIO_SET_DTR_MASK 0x1
#define SIO_SET_DTR_HIGH ( 1 | ( SIO_SET_DTR_MASK  << 8))
#define SIO_SET_DTR_LOW  ( 0 | ( SIO_SET_DTR_MASK  << 8))
#define SIO_SET_RTS_MASK 0x2
#define SIO_SET_RTS_HIGH ( 2 | ( SIO_SET_RTS_MASK << 8 ))
#define SIO_SET_RTS_LOW ( 0 | ( SIO_SET_RTS_MASK << 8 ))

#define SIO_RTS_CTS_HS (0x1 << 8)

/* marker for unused usb urb structures
   (taken from libusb) */
#define BRITEBLOX_URB_USERCONTEXT_COOKIE ((void *)0x1)

#ifdef __GNUC__
#define DEPRECATED(func) func __attribute__ ((deprecated))
#elif defined(_MSC_VER)
#define DEPRECATED(func) __declspec(deprecated) func
#else
#pragma message("WARNING: You need to implement DEPRECATED for this compiler")
#define DEPRECATED(func) func
#endif

struct briteblox_transfer_control
{
    int completed;
    unsigned char *buf;
    int size;
    int offset;
    struct briteblox_context *briteblox;
    struct libusb_transfer *transfer;
};

/**
    \brief Main context structure for all libbriteblox functions.

    Do not access directly if possible.
*/
struct briteblox_context
{
    /* USB specific */
    /** libusb's context */
    struct libusb_context *usb_ctx;
    /** libusb's usb_dev_handle */
    struct libusb_device_handle *usb_dev;
    /** usb read timeout */
    int usb_read_timeout;
    /** usb write timeout */
    int usb_write_timeout;

    /* BRITEBLOX specific */
    /** BRITEBLOX chip type */
    enum briteblox_chip_type type;
    /** baudrate */
    int baudrate;
    /** bitbang mode state */
    unsigned char bitbang_enabled;
    /** pointer to read buffer for briteblox_read_data */
    unsigned char *readbuffer;
    /** read buffer offset */
    unsigned int readbuffer_offset;
    /** number of remaining data in internal read buffer */
    unsigned int readbuffer_remaining;
    /** read buffer chunk size */
    unsigned int readbuffer_chunksize;
    /** write buffer chunk size */
    unsigned int writebuffer_chunksize;
    /** maximum packet size. Needed for filtering modem status bytes every n packets. */
    unsigned int max_packet_size;

    /* BRITEBLOX FT2232C requirecments */
    /** FT2232C interface number: 0 or 1 */
    int interface;   /* 0 or 1 */
    /** FT2232C index number: 1 or 2 */
    int index;       /* 1 or 2 */
    /* Endpoints */
    /** FT2232C end points: 1 or 2 */
    int in_ep;
    int out_ep;      /* 1 or 2 */

    /** Bitbang mode. 1: (default) Normal bitbang mode, 2: FT2232C SPI bitbang mode */
    unsigned char bitbang_mode;

    /** Decoded eeprom structure */
    struct briteblox_eeprom *eeprom;

    /** String representation of last error */
    char *error_str;

    /** Defines behavior in case a kernel module is already attached to the device */
    enum briteblox_module_detach_mode module_detach_mode;
};

/**
 List all handled EEPROM values.
   Append future new values only at the end to provide API/ABI stability*/
enum briteblox_eeprom_value
{
    VENDOR_ID          = 0,
    PRODUCT_ID         = 1,
    SELF_POWERED       = 2,
    REMOTE_WAKEUP      = 3,
    IS_NOT_PNP         = 4,
    SUSPEND_DBUS7      = 5,
    IN_IS_ISOCHRONOUS  = 6,
    OUT_IS_ISOCHRONOUS = 7,
    SUSPEND_PULL_DOWNS = 8,
    USE_SERIAL         = 9,
    USB_VERSION        = 10,
    USE_USB_VERSION    = 11,
    MAX_POWER          = 12,
    CHANNEL_A_TYPE     = 13,
    CHANNEL_B_TYPE     = 14,
    CHANNEL_A_DRIVER   = 15,
    CHANNEL_B_DRIVER   = 16,
    CBUS_FUNCTION_0    = 17,
    CBUS_FUNCTION_1    = 18,
    CBUS_FUNCTION_2    = 19,
    CBUS_FUNCTION_3    = 20,
    CBUS_FUNCTION_4    = 21,
    CBUS_FUNCTION_5    = 22,
    CBUS_FUNCTION_6    = 23,
    CBUS_FUNCTION_7    = 24,
    CBUS_FUNCTION_8    = 25,
    CBUS_FUNCTION_9    = 26,
    HIGH_CURRENT       = 27,
    HIGH_CURRENT_A     = 28,
    HIGH_CURRENT_B     = 29,
    INVERT             = 30,
    GROUP0_DRIVE       = 31,
    GROUP0_SCHMITT     = 32,
    GROUP0_SLEW        = 33,
    GROUP1_DRIVE       = 34,
    GROUP1_SCHMITT     = 35,
    GROUP1_SLEW        = 36,
    GROUP2_DRIVE       = 37,
    GROUP2_SCHMITT     = 38,
    GROUP2_SLEW        = 39,
    GROUP3_DRIVE       = 40,
    GROUP3_SCHMITT     = 41,
    GROUP3_SLEW        = 42,
    CHIP_SIZE          = 43,
    CHIP_TYPE          = 44,
    POWER_SAVE         = 45,
    CLOCK_POLARITY     = 46,
    DATA_ORDER         = 47,
    FLOW_CONTROL       = 48,
    CHANNEL_C_DRIVER   = 49,
    CHANNEL_D_DRIVER   = 50,
    CHANNEL_A_RS485    = 51,
    CHANNEL_B_RS485    = 52,
    CHANNEL_C_RS485    = 53,
    CHANNEL_D_RS485    = 54,
    RELEASE_NUMBER     = 55,
};

/**
    \brief list of usb devices created by briteblox_usb_find_all()
*/
struct briteblox_device_list
{
    /** pointer to next entry */
    struct briteblox_device_list *next;
    /** pointer to libusb's usb_device */
    struct libusb_device *dev;
};
#define FT1284_CLK_IDLE_STATE 0x01
#define FT1284_DATA_LSB       0x02 /* DS_FT232H 1.3 amd ftd2xx.h 1.0.4 disagree here*/
#define FT1284_FLOW_CONTROL   0x04
#define POWER_SAVE_DISABLE_H 0x80

#define USE_SERIAL_NUM 0x08
enum briteblox_cbus_func  /* FIXME: Recheck value, especially the last */
{
    CBUS_TXDEN = 0, CBUS_PWREN = 1, CBUS_RXLED = 2, CBUS_TXLED = 3, CBUS_TXRXLED = 4,
    CBUS_SLEEP = 5, CBUS_CLK48 = 6, CBUS_CLK24 = 7, CBUS_CLK12 = 8, CBUS_CLK6 =  9,
    CBUS_IOMODE = 0xa, CBUS_BB_WR = 0xb, CBUS_BB_RD = 0xc, CBUS_BB   = 0xd
};

enum briteblox_cbush_func  /* FIXME: Recheck value, especially the last */
{
    CBUSH_TRISTATE = 0, CBUSH_RXLED = 1, CBUSH_TXLED = 2, CBUSH_TXRXLED = 3, CBUSH_PWREN = 4,
    CBUSH_SLEEP = 5, CBUSH_DRIVE_0 = 6, CBUSG_DRIVE1 = 7, CBUSH_IOMODE = 8, CBUSH_TXDEN =  9,
    CBUSH_CLK30 = 10, CBUSH_CLK15 = 11, CBUSH_CLK7_5 = 12, CBUSH_BAT_DETECT = 13,
    CBUSH_BAT_DETECT_NEG = 14, CBUSH_I2C_TXE = 15, CBUSH_I2C_RXF = 16, CBUSH_VBUS_SENSE = 17,
    CBUSH_BB_WR = 18, CBUSH_BB_RD = 19, CBUSH_TIME_STAMP = 20, CBUSH_AWAKE = 21,
};

/** Invert TXD# */
#define INVERT_TXD 0x01
/** Invert RXD# */
#define INVERT_RXD 0x02
/** Invert RTS# */
#define INVERT_RTS 0x04
/** Invert CTS# */
#define INVERT_CTS 0x08
/** Invert DTR# */
#define INVERT_DTR 0x10
/** Invert DSR# */
#define INVERT_DSR 0x20
/** Invert DCD# */
#define INVERT_DCD 0x40
/** Invert RI# */
#define INVERT_RI  0x80

/** Interface Mode. */
#define CHANNEL_IS_UART 0x0
#define CHANNEL_IS_FIFO 0x1
#define CHANNEL_IS_OPTO 0x2
#define CHANNEL_IS_CPU  0x4
#define CHANNEL_IS_FT1284 0x8

#define CHANNEL_IS_RS485 0x10

#define DRIVE_4MA  0
#define DRIVE_8MA  1
#define DRIVE_12MA 2
#define DRIVE_16MA 3
#define SLOW_SLEW  4
#define IS_SCHMITT 8

/** Driver Type. */
#define DRIVER_VCP 0x08
#define DRIVER_VCPH 0x10 /* FT232H has moved the VCP bit */

#define USE_USB_VERSION_BIT 0x10

#define SUSPEND_DBUS7_BIT 0x80

/** High current drive. */
#define HIGH_CURRENT_DRIVE   0x10
#define HIGH_CURRENT_DRIVE_R 0x04

/**
    \brief Progress Info for streaming read
*/
struct size_and_time
{
    uint64_t totalBytes;
    struct timeval time;
};

typedef struct
{
    struct size_and_time first;
    struct size_and_time prev;
    struct size_and_time current;
    double totalTime;
    double totalRate;
    double currentRate;
} BRITEBLOXProgressInfo;

typedef int (BRITEBLOXStreamCallback)(uint8_t *buffer, int length,
                                 BRITEBLOXProgressInfo *progress, void *userdata);

/**
 * Provide libbriteblox version information
 * major: Library major version
 * minor: Library minor version
 * micro: Currently unused, ight get used for hotfixes.
 * version_str: Version as (static) string
 * snapshot_str: Git snapshot version if known. Otherwise "unknown" or empty string.
*/
struct briteblox_version_info
{
    int major;
    int minor;
    int micro;
    const char *version_str;
    const char *snapshot_str;
};


#ifdef __cplusplus
extern "C"
{
#endif

    int briteblox_init(struct briteblox_context *briteblox);
    struct briteblox_context *briteblox_new(void);
    int briteblox_set_interface(struct briteblox_context *briteblox, enum briteblox_interface interface);

    void briteblox_deinit(struct briteblox_context *briteblox);
    void briteblox_free(struct briteblox_context *briteblox);
    void briteblox_set_usbdev (struct briteblox_context *briteblox, struct libusb_device_handle *usbdev);

    struct briteblox_version_info briteblox_get_library_version(void);

    int briteblox_usb_find_all(struct briteblox_context *briteblox, struct briteblox_device_list **devlist,
                          int vendor, int product);
    void briteblox_list_free(struct briteblox_device_list **devlist);
    void briteblox_list_free2(struct briteblox_device_list *devlist);
    int briteblox_usb_get_strings(struct briteblox_context *briteblox, struct libusb_device *dev,
                             char * manufacturer, int mnf_len,
                             char * description, int desc_len,
                             char * serial, int serial_len);
    int briteblox_eeprom_set_strings(struct briteblox_context *briteblox, char * manufacturer,
                                char * product, char * serial);

    int briteblox_usb_open(struct briteblox_context *briteblox, int vendor, int product);
    int briteblox_usb_open_desc(struct briteblox_context *briteblox, int vendor, int product,
                           const char* description, const char* serial);
    int briteblox_usb_open_desc_index(struct briteblox_context *briteblox, int vendor, int product,
                                 const char* description, const char* serial, unsigned int index);
    int briteblox_usb_open_dev(struct briteblox_context *briteblox, struct libusb_device *dev);
    int briteblox_usb_open_string(struct briteblox_context *briteblox, const char* description);

    int briteblox_usb_close(struct briteblox_context *briteblox);
    int briteblox_usb_reset(struct briteblox_context *briteblox);
    int briteblox_usb_purge_rx_buffer(struct briteblox_context *briteblox);
    int briteblox_usb_purge_tx_buffer(struct briteblox_context *briteblox);
    int briteblox_usb_purge_buffers(struct briteblox_context *briteblox);

    int briteblox_set_baudrate(struct briteblox_context *briteblox, int baudrate);
    int briteblox_set_line_property(struct briteblox_context *briteblox, enum briteblox_bits_type bits,
                               enum briteblox_stopbits_type sbit, enum briteblox_parity_type parity);
    int briteblox_set_line_property2(struct briteblox_context *briteblox, enum briteblox_bits_type bits,
                                enum briteblox_stopbits_type sbit, enum briteblox_parity_type parity,
                                enum briteblox_break_type break_type);

    int briteblox_read_data(struct briteblox_context *briteblox, unsigned char *buf, int size);
    int briteblox_read_data_set_chunksize(struct briteblox_context *briteblox, unsigned int chunksize);
    int briteblox_read_data_get_chunksize(struct briteblox_context *briteblox, unsigned int *chunksize);

    int briteblox_write_data(struct briteblox_context *briteblox, const unsigned char *buf, int size);
    int briteblox_write_data_set_chunksize(struct briteblox_context *briteblox, unsigned int chunksize);
    int briteblox_write_data_get_chunksize(struct briteblox_context *briteblox, unsigned int *chunksize);

    int briteblox_readstream(struct briteblox_context *briteblox, BRITEBLOXStreamCallback *callback,
                        void *userdata, int packetsPerTransfer, int numTransfers);
    struct briteblox_transfer_control *briteblox_write_data_submit(struct briteblox_context *briteblox, unsigned char *buf, int size);

    struct briteblox_transfer_control *briteblox_read_data_submit(struct briteblox_context *briteblox, unsigned char *buf, int size);
    int briteblox_transfer_data_done(struct briteblox_transfer_control *tc);

    int briteblox_set_bitmode(struct briteblox_context *briteblox, unsigned char bitmask, unsigned char mode);
    int briteblox_disable_bitbang(struct briteblox_context *briteblox);
    int briteblox_read_pins(struct briteblox_context *briteblox, unsigned char *pins);

    int briteblox_set_latency_timer(struct briteblox_context *briteblox, unsigned char latency);
    int briteblox_get_latency_timer(struct briteblox_context *briteblox, unsigned char *latency);

    int briteblox_poll_modem_status(struct briteblox_context *briteblox, unsigned short *status);

    /* flow control */
    int briteblox_setflowctrl(struct briteblox_context *briteblox, int flowctrl);
    int briteblox_setdtr_rts(struct briteblox_context *briteblox, int dtr, int rts);
    int briteblox_setdtr(struct briteblox_context *briteblox, int state);
    int briteblox_setrts(struct briteblox_context *briteblox, int state);

    int briteblox_set_event_char(struct briteblox_context *briteblox, unsigned char eventch, unsigned char enable);
    int briteblox_set_error_char(struct briteblox_context *briteblox, unsigned char errorch, unsigned char enable);

    /* init eeprom for the given BRITEBLOX type */
    int briteblox_eeprom_initdefaults(struct briteblox_context *briteblox,
                                 char * manufacturer, char *product,
                                 char * serial);
    int briteblox_eeprom_build(struct briteblox_context *briteblox);
    int briteblox_eeprom_decode(struct briteblox_context *briteblox, int verbose);

    int briteblox_get_eeprom_value(struct briteblox_context *briteblox, enum briteblox_eeprom_value value_name, int* value);
    int briteblox_set_eeprom_value(struct briteblox_context *briteblox, enum briteblox_eeprom_value value_name, int  value);

    int briteblox_get_eeprom_buf(struct briteblox_context *briteblox, unsigned char * buf, int size);
    int briteblox_set_eeprom_buf(struct briteblox_context *briteblox, const unsigned char * buf, int size);

    int briteblox_read_eeprom(struct briteblox_context *briteblox);
    int briteblox_read_chipid(struct briteblox_context *briteblox, unsigned int *chipid);
    int briteblox_write_eeprom(struct briteblox_context *briteblox);
    int briteblox_erase_eeprom(struct briteblox_context *briteblox);

    int briteblox_read_eeprom_location (struct briteblox_context *briteblox, int eeprom_addr, unsigned short *eeprom_val);
    int briteblox_write_eeprom_location(struct briteblox_context *briteblox, int eeprom_addr, unsigned short eeprom_val);

    char *briteblox_get_error_string(struct briteblox_context *briteblox);

#ifdef __cplusplus
}
#endif

#endif /* __libbriteblox_h__ */
